### ISSUE TITLE
Concurrent poll&offer generates NPE

### ISSUE THREAD

**Issue by zysaaa** [ben-manes/lock-free-queue#1]
Hello, I understand that this is a long time ago project, I just want to learn...  I found that ```NPE``` exception will be generated, see the following code:

```
public static void main(String[] args) throws InterruptedException {
        ConcurrentSingleConsumerQueue<Integer> queue = new ConcurrentSingleConsumerQueue<>(10);
        int num = 1000;
        new Thread(() -> {
            int k = num;
            while (k != 0) {
                try {
                    if (queue.poll() != null) {
                        k--;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    break;
                }
            }
        }).start();
        for (int i = 0; i < num; i++) {
            queue.offer(i);
        }
    }
```
It seems that the update(in ```offer```) and read(in ```poll```) of variable  ```tail``` in ```ConcurrentSingleConsumerQueue``` will cause this bug:

```
    @Override
    public boolean offer(E e) {
        if (e == null) {
            throw new NullPointerException();
        }
        long t = tail.getAndIncrement();
        long h = head.get();
        // here, tail already update, but data hasn't been inserted yet
        // sleep 1 s here, high probability of recurrence this bug
        try {
            Thread.sleep(1000);
        } catch (InterruptedException ite) {}
        if ((t - h) < array.length()) {
            int index = (int) (t & mask);
            array.lazySet(index, e);
        } else {
            Node<E> node = new Node<E>(e);
            tailNode.getAndSet(node).lazySet(node);
        }
        return true;
    }
```

**Comment by ben-manes** [comment]
I don't remember the code, so I presumably abandoned it before it was thread-safe.

I did write a [later version](https://github.com/ben-manes/caffeine/blob/v2.dev/caffeine/src/main/java/com/github/benmanes/caffeine/SingleConsumerQueue.java) that was used by the caching library. That passes your test case. It was replaced with a [JCTools](https://github.com/JCTools/JCTools) `MpscGrowableArrayQueue`, which is array-based rather than linked-list and therefore can have better GC behavior (as entries won't cross the young/old gen boundaries). I think you'll find jctools a better code base to learn from.
